### PR TITLE
🐛 Source Shopify: Fixed rate limits throttling

### LINKED ISSUE #5416: source shopify: confusing error message when rate limit is hit
This is the image a user saw in their logs when the shopify connector hit rate limits: 

image

we should display a clearer error message

### LINKED ISSUE #5441: Shopify Connector Rate Limiting
## Enviroment
- **Airbyte version**:  example is 0.27.1-alpha
- **OS Version / Instance**:  AWS EC2
- **Deployment**: Docker
- **Source Connector and version**: Shopify 0.1.11
- **Destination Connector and version**: Postgres
- **Severity**: High
- **Step where error happened**: Sync job

## Current Behavior
Right now, incremental sync jobs are taking somewhere between 12-14 hours. Given the size of the data (and performance we've seen with other ETL providers), I think the way in which data is being fetched may be suboptimal. It definitely could be the case that jobs do indeed require this amount of time, but am seeing loads for ~1 hour of syncing/backing off for with at most 100s of records, which should be well below the rate limit given that much time. 

Overall, 1,017 records read took 13 hours and 55 minutes on our most recent sync. According to the [Shopify API](https://shopify.dev/api/usage/rate-limits), we should be able to do 2 requests/second/store (with leaky bucket approach), which should be well below for that many records loaded.

All of the tables in this sync are being replicated using the "Incremental - Deduped + History" sync strategy.

## Expected Behavior
Load time should be <1 hour given Shopify rate limiting strategy.

## Logs
*If applicable, please upload the logs from the failing operation. 
For sync jobs, you can download the full logs from the UI by going to the sync attempt page and 
clicking the download logs button at the top right of the logs display window.*

(Logs attached)
[logs-39-0.txt](https://github.com/airbytehq/airbyte/files/6995320/logs-39-0.txt)

<details>
<summary>LOG</summary> 

```
```

</details>

## Steps to Reproduce
1. Create a Shopify source/destination
2. Run sync jobs on a 24 hour schedule
3. Monitor sync times

## Are you willing to submit a PR?

Let me know if helpful

### PR DESCRIPTION
**Pull request by bazarnov**
## What
#5441 - Shopify Connector Rate Limiting
#5416 - Source shopify: confusing error message when rate limit is hit

## How
- added decorator `balance_rate_limit` method to control sleep time between each API call, by leveraging the `"X-Shopify-Shop-Api-Call-Limit"` header information, as described in [Shopify Documentation](https://shopify.dev/api/usage/rate-limits)
- added unit tests to cover the issue
- added note in `Performance considerations` section of `shopify.md` about  rate limits.

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
 
- [X] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [X] Documentation updated 
    - [X] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [X] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [X] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py
Comment (antixar) [inline airbyte-integrations/connectors/source-shopify/source_shopify/source.py:106]: Typo: t**h**reshold

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py
Comment (antixar) [inline airbyte-integrations/connectors/source-shopify/source_shopify/source.py:93]: add an example of values for this header

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py
Comment (antixar) [inline airbyte-integrations/connectors/source-shopify/source_shopify/source.py:113]: Maybe would implementation of this delay as a function decorator be better and obvious?

File: airbyte-integrations/connectors/source-shopify/unit_tests/unit_test.py
Comment (antixar) [inline airbyte-integrations/connectors/source-shopify/unit_tests/unit_test.py:45]: Can we use these link values from streams' structs?

Comment (antixar) [reply airbyte-integrations/connectors/source-shopify/unit_tests/unit_test.py:45]: Also same point is and for other tests.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:105]: Fixed.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:92]: Done.

File: airbyte-integrations/connectors/source-shopify/unit_tests/unit_test.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/unit_tests/unit_test.py:45]: I believe there is no reason for this, since this is the mock) but if you insist - I'll do my best.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:112]: Done.

**Review by annalvova05 — APPROVED** [review]

File: airbyte-integrations/connectors/source-shopify/unit_tests/test_control_rate_limit.py
Comment (antixar) [inline airbyte-integrations/connectors/source-shopify/unit_tests/test_control_rate_limit.py:37]: Use capital letters  for global variables ([pep8](https://www.python.org/dev/peps/pep-0008/#constants))

**Review by antixar — APPROVED** [review]

File: docs/integrations/sources/shopify.md
Comment (Phlair) [inline docs/integrations/sources/shopify.md:48]:
```suggestion
Typically, there should not be issues with throttling or exceeding the rate limits but in some edge cases, user can receive the warning message as follows:
```

File: docs/integrations/sources/shopify.md
Comment (Phlair) [inline docs/integrations/sources/shopify.md:53]:
```suggestion
This is expected when the connector hits the 429 - Rate Limit Exceeded HTTP Error.
With given error message the sync operation still goes on, but will require more time to finish.
```

**Review by Phlair — COMMENTED** [review]
Like the decorator for throttling sleeps and thanks for adding unit tests.

Couple of comments and Qs :)

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/utils.py:62]: Done.

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/utils.py:78]: > Is this ever expected to happen?
Yes. There is a stream `abbandoned_checkouts` that doesn't have the Rate-Limits Header, thus we wait the Shopify recommended time between each request, which is 1 sec. There should not be any issues waiting 1 sec - it's Shopify recommended.

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/utils.py:59]: Ok.

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/utils.py:81]: You're right) Somehow missed this, my bad. Corrected.

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/utils.py:53]: There is only 1 Shopify recommendation for general "less than rate-limit" data querying:
`2 requests/sec`...."Make requests with 1 sec delay". 

However this approach is a bit poor, because why should we wait 1 sec between each request if we have the capacity not to, and if we hit the approximate mid_load then we slow down and use some delay, once we have more capacity again - we use full burst. 
So, 
the 0.2 is the standard delay found experimentally - it's a perfect ratio between delay and request,
the 1.5 - is a compromise between recommended wait time and delay, where we wait 2 seconds to have another 4 requests capacity.
1.0 - is the standard recommended wait_time - we use it only if we don't have any `rate_limit` header.
5.0 - is the maximum time we wait in order to regain capacity back, each 1 sec of wait time gives us another 4 requests capacity, thus after 5 seconds of waiting without catch 429 HTTP Error gives us perfect rate_limit recovery and we have almost 90% of capacity back. We can always wait for 2.0 sec, as Retry-After provides us, but again, we have it only after 429 Error, and it's not sufficient, because should try to never catch that error.

File: docs/integrations/sources/shopify.md
Comment (bazarnov) [reply docs/integrations/sources/shopify.md:48]: Corrected.

File: docs/integrations/sources/shopify.md
Comment (bazarnov) [reply docs/integrations/sources/shopify.md:53]: Corrected.

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/utils.py:40]: Added.

File: airbyte-integrations/connectors/source-shopify/unit_tests/test_control_rate_limit.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/unit_tests/test_control_rate_limit.py]: Done.

File: airbyte-integrations/connectors/source-shopify/unit_tests/test_control_rate_limit.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/unit_tests/test_control_rate_limit.py]: Done

**Comment by rparrapy** [comment]
FWIW, I am eagerly waiting for this to be merged :)

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py
Comment (Phlair) [reply airbyte-integrations/connectors/source-shopify/source_shopify/utils.py]: Thanks for detailed explanation, would be a nice to have that in the code as comment (or maybe the README.md?)

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py
Comment (Phlair) [inline airbyte-integrations/connectors/source-shopify/source_shopify/utils.py:31]: I'm not sure I see the reason for three separate classes here.
I was thinking you could just have one class which holds:
- the "on_x_load" attributes at class level
- the `get_wait_time` func
- the decorator func

**Review by Phlair — APPROVED** [review]
Thanks for making changes!
I'm approving now so we can get this released for users but for readability and structure sake, addressing the comment I left would be ideal. If not then at least making an issue to do so would be great (or let me know if you disagree!)

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/utils.py:31]: Done.

**Pull request merged**
